File: no-internet/unique-paths-ii.ts
/**
 * https://leetcode.com/problems/unique-paths-ii/
 */
export { uniquePathsWithObstacles };

function uniquePathsWithObstacles(obstacleGrid: number[][]): number {
  function dfs(x: number, y: number): number {
    const lastX = obstacleGrid.length - 1;
    const lastY = obstacleGrid[0].length - 1;
    if (x > lastX || y > lastY) return 0;
    if (obstacleGrid[x][y] === 1) return 0;
    if (x === lastX && y === lastY) return 1;

    let sum = 0;
    sum += dfs(x + 1, y);
    sum += dfs(x, y + 1);
    return sum;
  }

  return dfs(0, 0);
}

File: python/today4/unique_paths_ii.py
# https://leetcode.com/problems/unique-paths-ii/description/
from typing import Dict, List

class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        last_row = len(obstacleGrid) - 1
        last_col = len(obstacleGrid[0]) - 1
        memo: Dict[str, int] = {}
        
        def dfs(row: int, col: int) -> int:
            if row > last_row or col > last_col or obstacleGrid[row][col] == 1:
                return 0
            if row == last_row and col == last_col:
                return 1
            key = str(row) + '-' + str(col)
            if key in memo:
                return memo[key]
            
            count = dfs(row, col+1)
            count += dfs(row+1, col)

            memo[key] = count
            return count

        return dfs(0, 0)